R/utilities-dimensions.R

```diff
@@ -0,0 +1,162 @@
+  if (!existsDimension(dimension)) {
+    stop(messages$errorDimensionNotSupported(dimension))
+  }
+}
+
+#' Unit existence
+#'
+#' @param unit String name of the unit
+#' @param dimension String name of the dimension.
+#' @details Check if the unit is valid for the dimension.
+#' @export
+hasUnit <- function(unit, dimension) {
+  validateIsString(c(unit, dimension))
+  validateDimension(dimension)
+  dimensionTask <- getDimensionTask()
+  # Get the dimension object
+  dimension <- rClr::clrCall(dimensionTask, "DimensionByName", enc2utf8(dimension))
+  rClr::clrCall(dimension, "HasUnit", enc2utf8(unit))
+}
+
+#' Validate unit
+#'
+#' @param unit String name of the unit
+#' @param dimension String name of the dimension.
+#' @details Check if the unit is valid for the dimension. If not, throw an error
+#' @export
+validateUnit <- function(unit, dimension) {
+  if (!hasUnit(unit, dimension)) {
+    stop(messages$errorUnitNotSupported(unit, dimension))
+  }
+}
+
+#' Get base unit of a dimension
+#'
+#' @param dimension Dimension for which the base unit is returned.
+#'
+#' @return String name of the base unit.
+#' @export
+getBaseUnit <- function(dimension) {
+  validateDimension(dimension)
+  dimensionTask <- getDimensionTask()
+  # Get the dimension object
+  dimension <- rClr::clrCall(dimensionTask, "DimensionByName", enc2utf8(dimension))
+  unit <- rClr::clrCall(dimension, "get_BaseUnit")
+  rClr::clrCall(unit, "ToString")
+}
+
+#' Calculate factor for converting values of one dimension into another
+#'
+#' @param dimension1 String name of the first dimension
+#' @param dimension2 String name fo the second dimension
+#' @param mw Double value of the molecular weight of the molecule. Used
+#' for conversion between molar and mass concentrations or amounts. Can be \code{NULL}
+#' for other dimensions (default)
+#'
+#' @return A numerical factor the values of the first dimensions must be multiplied by to
+#' be converted to second dimension. The conversion assumes values being in base units
+#' of both dimensions.
+#' If the values cannot be transformed between the dimensions, an error is thrown.
+#' @export
+dimensionsConversionFactor <- function(dimension1, dimension2, mw = NULL) {
+  # No conversion if the dimensions are equall
+  validateIsString(c(dimension1, dimension2))
+  if (dimension1 == dimension2) {
+    return(1)
+  }
+
+  # Dimensionless <-> Fraction
+  if (dimension1 == Dimensions$Dimensionless) {
+    if (dimension2 == Dimensions$Fraction) {
+      return(1)
+    }
+  }
+  if (dimension1 == Dimensions$Fraction) {
+    if (dimension2 == Dimensions$Dimensionless) {
+      return(1)
+    }
+  }
+
+  # Concentration (molar) <-> Concentration (mass)
+  if (dimension1 == Dimensions$`Concentration (molar)`) {
+    if (dimension2 == Dimensions$`Concentration (mass)`) {
+      if (is.null(mw)) {
+        stop(messages$errorCannotConvertDimensionsNoMW(dimension1, dimension2))
+      }
+      return(1 * 1e-6 * mw * 1e-3)

```

Comment:
This is not something I want to see here. I need to understand the usage of such concept. We have already everything encapsulated in Core

Reply:
Would be happy to see another way. E.g. - I read data in mg/dl and want to convert it to mol/l. How? Without having a Quantity object.

Reply:
There are plenty of method exposted directly in core doing this. Creating a conversation factor from one dimension to another dimension does not make sense BTW. It needs to be from one unit to another unit

In core see

```
  double[] ConvertToUnit(string dimensionName, string targetUnit, double[] valuesInBaseUnit, double molWeight);
      double[] ConvertToUnit(string dimensionName, string targetUnit, double[] valuesInBaseUnit);

      double[] ConvertToUnit(string dimensionName, string targetUnit, double valueInBaseUnit, double molWeight);
      double[] ConvertToUnit(string dimensionName, string targetUnit, double valueInBaseUnit);

```

and 
```
 double[] ConvertToBaseUnit(string dimensionName, string displayUnit, double[] valuesInDisplayUnit, double molWeight);
      double[] ConvertToBaseUnit(string dimensionName, string displayUnit, double[] valuesInDisplayUnit);

      double[] ConvertToBaseUnit(string dimensionName, string displayUnit, double valueInDisplayUnit, double molWeight);
      double[] ConvertToBaseUnit(string dimensionName, string displayUnit, double valueInDisplayUnit);

```

All exposed via the DImensionTag. IT deals EXPLICITELY with those use cases